python/MOCK_sensors.py

```diff
@@ -122,8 +130,148 @@ def publish(self):
+
+            if(self.brokenSensorData is None):
+                randomSeed = get_rosparam_or_default_if_invalid('random_seed', default=None, rosparam_type_cast=str)
+                random.seed(randomSeed)
+                self.brokenSensorData = random.randint(1, 3)
+
+            self.corrupt_wind_sensor(data, self.brokenSensorData, 0, 0)
+
+            rospy.logwarn_throttle(3, "MOCK wind_sensor_{} publishing corrupt data: ".format(self.brokenSensorData)
+                                   + "angle and speed stuck at 0")
+
+        elif (brokenWindSensorStatus == "stuck_at_rand_constant"):
+            # This setting chooses a wind sensor which reads a constant random speed and direction
+
+            if(self.brokenSensorData is None):
+                randomSeed = get_rosparam_or_default_if_invalid('random_seed', default=None, rosparam_type_cast=str)
+                random.seed(randomSeed)
+
+                self.brokenSensorData = {'brokenSensorID': random.randint(1, 3),
+                                         'brokenSensorAngle_deg': random.random() * 360,
+                                         'brokenSensorSpeed_knots': random.random() * MAX_BROKEN_WINDSPEED_KNOTS}
+
+            self.corrupt_wind_sensor(data, self.brokenSensorData['brokenSensorID'],
+                                     self.brokenSensorData['brokenSensorAngle_deg'],
+                                     self.brokenSensorData['brokenSensorSpeed_knots'])
+
+            rospy.logwarn_throttle(3, "MOCK wind_sensor_{} ".format(self.brokenSensorData['brokenSensorID'])
+                                   + "publishing corrupt data:\n"
+                                   + "angle stuck at {} degrees,".format(self.brokenSensorData['brokenSensorAngle_deg'])
+                                   + " speed stuck at {}".format(self.brokenSensorData['brokenSensorSpeed_knots'])
+                                   + " knots.")
+
+        elif(brokenWindSensorStatus == "rand_const_error"):
+            # This setting chooses a wind sensor and corrupts its data as follows:
+            # the angle published is the original angle plus a random constant theta
+            # the speed published is given by a*(original speed) + b; a and b are random constants
+            # a is within [1 / MAX_WINDSPEED_MULTIPLIER_ERROR, MAX_WINDSPEED_MULTIPLIER_ERROR]
+            # b is within [-MAX_WINDSPEED_ADDER_ERROR_KNOTS, MAX_WINDSPEED_ADDER_ERROR_KNOTS]
+            # theta is within [-MAX_ANGLE_ERROR_DEG, MAX_ANGLE_ERROR_DEG]
+
+            if(self.brokenSensorData is None):
+                randomSeed = get_rosparam_or_default_if_invalid('random_seed', default=None, rosparam_type_cast=str)
+                random.seed(randomSeed)
+
+                if random.random < 0.5:
+                    multiplier = 1 - random.random() * (1 - 1 / (MAX_WINDSPEED_MULTIPLIER_ERROR))
+                else:
+                    multiplier = 1 + random.random() * (MAX_WINDSPEED_MULTIPLIER_ERROR - 1)
```

Comment:
Let's discuss this

Reply:
I wasn't satisfied with how I explained this code earlier, so I thought I'd give it another try after looking it over, and work through an example to check myself on my reasoning. Here is an example to show the behaviour I'm going for:

Suppose `MAX_WINDSPEED_MULTIPLIER_ERROR` is set to 1.2. I want the multiplier to be between 1/1.2 = 0.833 and 1.2. The first case is `1 - (a number between 0 and 1 - 1/1.2 = 0.167)`, which gives us a multiplier between 0.833 and 1. The second case is `1 + (a number between 0 and 0.2)`, which gives us a multiplier between 1 and 1.2 as desired.

It would have been more simple to read if the second case just said `muliplier = random.random()*MAX_WINDSPEED_MULTIPLIER_ERROR`, but that leaves the possibility of `random.random()` being close to zero which would lead to the multiplier being outside of the desired range from 0.833 to 1.2.

I could have instead chosen to create a multiplier between `1 - MAX_WINDSPEED_MULTIPLIER_ERROR` and `1 + MAX_WINDSPEED_MULTIPLIER_ERROR`. However, if `MAX_WINDSPEED_MULTIPLIER_ERROR` were set to 1, this would create a multiplier between 0 and 2, and reading 0 all the time seems like a much more severe error than reading double the true windspeed.

(also I realised I was running into issues with integer division, so I changed the 1's to 1.0's)

Reply:
Can you post this in a new comment so that it is easier to find? thanks